### PR TITLE
[5.8] Add `register` method to the ServiceProvider

### PR DESCRIPTION
**Pull request by TBlindaruk**
In the https://github.com/laravel/framework/pull/27067 pr was added DeferrableProvider service provider.
 This Pr will add pretty the same thing.
 In case, if it will be accepted then I will add documentation for this. Also will be good to remove calling just `register` method from the provider in case if the method exists in the 5.9 release .

The main goal of these changes, that it is not so transparent, that your service will be registered in case if you will just add this method.

So my proposition to call `register` method only in case. if the class will implement the interface (I have providers witch calles only the `boot` method )

----------------

Also i thing will be good to add diferent contract for the:
- method `boot`
- prop `bindings` and `singletons`

-----------------------

UPD: just add register method to the service provider

### CODE REVIEW

**Comment by xuanquynh** [comment]
The code will be more pretty than too many method_exists or property_exists checks.

File: src/Illuminate/Contracts/Support/RegistrableProvider.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Illuminate\Contracts\Support;
+
+interface RegistrableProvider
```
Comment (koenhoeijmakers) [inline src/Illuminate/Contracts/Support/RegistrableProvider.php]: Shouldn't it be `RegisterableProvider`?

**Review by koenhoeijmakers — CHANGES_REQUESTED** [review]

File: src/Illuminate/Contracts/Support/RegistrableProvider.php (hunk 1 above)
Comment (xuanquynh) [reply src/Illuminate/Contracts/Support/RegistrableProvider.php]: I think `RegistrableProvider` is correct.

**Comment by driesvints** [comment]
Hmm, I'm not sure how I feel about this. I like the solution (although not sure about the name) but at the same time it's adding a bit more complexity for setting up service providers.

**Comment by TBlindaruk** [comment]
As for me will be ok also just put an empty `register` method to the base service provider ^^. And it will be calls every time, but nothing will be happens in case if no code within the `register` method

**Comment by driesvints** [comment]
@TBlindaruk that's indeed a better solution imo, however I feel that that doesn't changes much from the current situation.

**Comment by mfn** [comment]
But doesn't a ServiceProder **require** a register method to be useful, at all? I.e. shouldn't there be one an `abstract function register` being defined on the `ServiceProvider` in the first place (instead of an additional interface)

**Comment by driesvints** [comment]
@mfn not necessarily. You can also use `boot` to bootstrap functionality without registering services.

**Comment by mfn** [comment]
Thanks for correcting me 👍

**Comment by TBlindaruk** [comment]
@driesvints , changed

**Pull request merged**